lib/services/local/embeddings.go

```diff
@@ -0,0 +1,178 @@
+// Copyright 2023 Gravitational, Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package local
+
+import (
+	"context"
+	"crypto/sha256"
+	"time"
+
+	"github.com/gravitational/trace"
+	"github.com/jonboulle/clockwork"
+	"github.com/sirupsen/logrus"
+
+	"github.com/gravitational/teleport/api/utils/retryutils"
+	"github.com/gravitational/teleport/lib/ai"
+	"github.com/gravitational/teleport/lib/backend"
+	"github.com/gravitational/teleport/lib/services"
+)
+
+// EmbeddingsService implements the services.Embeddings interface.
+type EmbeddingsService struct {
+	log    *logrus.Entry
+	jitter retryutils.Jitter
+	backend.Backend
+	client ai.Embedder
+	clock  clockwork.Clock
+}
+
+const (
+	embeddingsPrefix = "embeddings"
+	embeddingExpiry  = 30 * 24 * time.Hour // 30 days
+)
+
+// GetEmbedding looks up a single embedding by its name in the backend.
+func (e EmbeddingsService) GetEmbedding(ctx context.Context, kind, resourceID string) (*ai.Embedding, error) {
+	result, err := e.Get(ctx, backend.Key(embeddingsPrefix, kind, resourceID))
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return services.UnmarshalEmbedding(result.Value)
+}
+
+// GetEmbeddings returns all embeddings for a given kind.
+func (e EmbeddingsService) GetEmbeddings(ctx context.Context, kind string) ([]*ai.Embedding, error) {
+	startKey := backend.Key(embeddingsPrefix, kind)
+	result, err := e.GetRange(ctx, startKey, backend.RangeEnd(startKey), backend.NoLimit)
```

Comment:
This will OOM the auth in a large enough cluster when asking for embeddings for `node`, since we intend to have (at least!) one embedding per active node.

Reply:
I switched to a stream. This function is currently called only in tests so it did not required large logic changes. If/when we implement garbage collection we'll be able to use the stream in a safe way.